astropy/coordinates/distances.py

```diff
@@ -142,13 +145,28 @@ def __new__(cls, value=None, unit=None, z=None, cosmology=None,
                 copy = False
 
             elif parallax is not None:
-                value = parallax.to(u.pc, equivalencies=u.parallax()).value
+                value = parallax.to_value(u.pc, equivalencies=u.parallax())
                 unit = u.pc
 
                 # Continue on to take account of unit and other arguments
                 # but a copy is already made, so no longer necessary
                 copy = False
 
+                if np.any(parallax < 0) and not allow_negative:
+                    raise ValueError('Negative parallaxes are not interpretable'
+                                     ' as distances. If you want to have these '
+                                     'values turned into NaN values, use the '
+                                     '`allow_negative=True` argument.')
+
+                elif np.any(parallax < 0) and allow_negative:
```

Comment:
Not to hold this back, but it seems performance wise that you could have a single ``if np.any(parallax < 0)`` then have ``if allow_negative:... else: ...`` inside (to avoid computing ``parallax < 0`` twice.

Reply:
Yep that's fair!